lightning/src/events/bump_transaction.rs

```diff
@@ -0,0 +1,597 @@
+	/// HTLC transaction. To sign HTLC inputs, an [`InMemorySigner`] should be re-derived through
+	/// [`KeysManager::derive_channel_keys`] with the help of `channel_keys_id` and
+	/// `channel_value_satoshis`. Each HTLC input's signature can be computed with
+	/// [`EcdsaChannelSigner::sign_holder_htlc_transaction`], which can then be provided to
+	/// [`HTLCDescriptor::tx_input_witness`] to obtain the fully signed witness required to spend.
+	///
+	/// It is possible to receive more than one instance of this event if a valid HTLC transaction
+	/// is never broadcast or is but not with a sufficient fee to be mined. Care should be taken by
+	/// the consumer of the event to ensure any future iterations of the HTLC transaction adhere to
+	/// the [Replace-By-Fee
+	/// rules](https://github.com/bitcoin/bitcoin/blob/master/doc/policy/mempool-replacements.md)
+	/// for fee bumps to be accepted into the mempool, and eventually the chain. As the frequency of
+	/// these events is not user-controlled, users may ignore/drop the event if either they are no
+	/// longer able to commit external confirmed funds to the HTLC transaction or the fee committed
+	/// to the HTLC transaction is greater in value than the HTLCs being claimed.
+	///
+	/// [`InMemorySigner`]: crate::chain::keysinterface::InMemorySigner
+	/// [`KeysManager::derive_channel_keys`]: crate::chain::keysinterface::KeysManager::derive_channel_keys
+	/// [`EcdsaChannelSigner::sign_holder_htlc_transaction`]: crate::chain::keysinterface::EcdsaChannelSigner::sign_holder_htlc_transaction
+	/// [`HTLCDescriptor::tx_input_witness`]: HTLCDescriptor::tx_input_witness
+	HTLCResolution {
+		/// The target feerate that the resulting HTLC transaction must meet.
+		target_feerate_sat_per_1000_weight: u32,
+		/// The set of pending HTLCs on the confirmed commitment that need to be claimed, preferably
+		/// by the same transaction.
+		htlc_descriptors: Vec<HTLCDescriptor>,
+		tx_lock_time: PackedLockTime,
+	},
+}
+
+pub struct Input {
+	pub outpoint: OutPoint,
+	pub witness_weight: u64,
+}
+
+#[derive(Clone)]
+pub struct Utxo {
+	pub outpoint: OutPoint,
+	pub output: TxOut,
+	pub witness_weight: u64,
+}
+
+pub struct CoinSelection {
+	confirmed_utxos: Vec<Utxo>,
+	fee_amount_sat: u64,
+	additional_tx_outputs: Vec<TxOut>,
+}
+
+pub trait CoinSelectionSource {
+	fn select_confirmed_utxos(
+		&self, must_spend: Vec<Input>, must_pay_to: Vec<TxOut>, target_feerate_sat_per_1000_weight: u32,
+	) -> Result<CoinSelection, ()>;
+	fn change_script(&self) -> Result<Script, ()>;
+	fn sign_confirmed_utxo(&self, tx: &Transaction, input: usize) -> Result<Witness, ()>;
+}
+
+pub trait WalletSource {
+	fn list_confirmed_utxos(&self) -> Result<Vec<Utxo>, ()>;
+	fn change_script(&self) -> Result<Script, ()>;
+	fn sign_confirmed_utxo(&self, tx: &Transaction, input: usize) -> Result<Witness, ()>;
+}
+
+fn select_confirmed_utxos<W: Deref>(
```

Comment:
Forgot to add this question that came up from review club: I'm imagining if someone's building a wallet or routing node or something that they might want to use something more optimized and would probably have access to bitcoin core or BDK to do so? Is the default implementation here meant to be used in real applications or more to make prototyping started easier (or both)?

Also if it's likely to be used in real applications, is it worth considering trying to provide the same default coin selection algorithm that BDK (and is one of three used in Bitcoin Core I believe) uses, i.e. [branch and bound](https://docs.rs/bdk/latest/bdk/wallet/coin_selection/type.DefaultCoinSelectionAlgorithm.html)?